src/core/thread/link_metrics.cpp

```diff
@@ -0,0 +1,353 @@
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for Thread Link Metrics.
+ */
+
+#include "link_metrics.hpp"
+
+#if OPENTHREAD_CONFIG_MLE_LINK_METRICS_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/instance.hpp"
+#include "common/locator-getters.hpp"
+#include "common/logging.hpp"
+
+#include "link_metrics_tlvs.hpp"
+
+namespace ot {
+
+LinkMetrics::LinkMetrics(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mLinkMetricsReportCallback(nullptr)
+    , mLinkMetricsReportCallbackContext(nullptr)
+{
+}
+
+otError LinkMetrics::LinkMetricsQuery(const Ip6::Address & aDestination,
+                                      uint8_t              aSeriesId,
+                                      const otLinkMetrics &aLinkMetricsFlags)
+{
+    otError                error;
+    LinkMetricsTypeIdFlags typeIdFlags[kMaxTypeIdFlags];
+    uint8_t                typeIdFlagsCount = GetTypeIdFlagsFromOtLinkMetricsFlags(typeIdFlags, aLinkMetricsFlags);
+
+    error = SendLinkMetricsQuery(aDestination, aSeriesId, typeIdFlags, typeIdFlagsCount);
+
+    return error;
+}
+
+otError LinkMetrics::AppendLinkMetricsReport(Message &                       aMessage,
+                                             const Mle::LinkMetricsQueryTlv *aLinkMetricsQuery,
+                                             const Message &                 aRequestMessage)
+{
+    otError                        error = OT_ERROR_NONE;
+    Tlv                            tlv;
+    const LinkMetricsQueryId *     queryId      = nullptr;
+    const LinkMetricsQueryOptions *queryOptions = nullptr;
+    uint8_t                        length       = 0;
+    uint16_t                       startOffset  = aMessage.GetLength();
+    const Tlv *                    subTlvCur    = nullptr;
+    const Tlv *                    subTlvEnd    = nullptr;
+
+    VerifyOrExit(aLinkMetricsQuery != nullptr && aLinkMetricsQuery->IsValid(), error = OT_ERROR_PARSE);
+    subTlvCur = aLinkMetricsQuery->GetSubTlvs();
+    subTlvEnd = aLinkMetricsQuery->GetNext();
+
```

Comment:
> Parse the ones we know already - keep track of what sub-TLVs we saw and parsed (bool variable to track)

Shall we directly use pointers (initialized to be `nullptr`) for possible TLVs? In this way we can directly use the tlvs after parsing. In `Leader::ValidatePrefix` the tlvs parsed are directly used as it's for validation.

Reply:
I think pointer would be tricky. Here is what I would suggest:
- We cannot read the entire `LinkMetricsQueryTlv` from message, since we cannot know its size (its value can be a bunch of sub-tlvs with different lengths, and we can/should not assume some max size for it)
- So in `MleRouter::SendDataResponse()` instead of using `Tlv::FindTlv()` (which would read the entire tlv from message into some buffer), we can use `FindTlvValueOffset()` to find offset of start of value portion and its length (or can use `FindTlvOffset()` to find the offset of start of the TLV).
- We can then pass this info (the offset into start of the value portion of `LinkMetricsQueryTlv` and its length)  in the call to `AppendLinkMetricsReport()`  (note that we already pass in the `aRequestMessage`)
- The value portion of `LinkMetricsQueryTlv` is a sequence of "Thread" Link Metric sub-TLVs.
- Then in `AppendLinkMetricsReport()` we try to parse and read all sub-tlvs from the message (knowing the start offset):
   - Read the TLV from `aRequestMessage` as a raw `Tlv tlv`.
   - In a `switch`, we can check the `tlv` type and if the type is one the ones we care about, we process it further and either read its value or read it fully again from message  (example of something similar to this is [`NetworkDiagnostic::GetNextDiagTlv()`](https://github.com/openthread/openthread/blob/master/src/core/thread/network_diagnostic.cpp#L763))
   - For simple TLVs with single value we don't even need to define a sub-class of `Tlv` type.
   - We can use helpers `Tlv::ReadUint8Tlv` or `ReadUint16Tlv`, etc to read the value directly from the message.
   - So for example we may not even need the class `LinkMetricsQueryId` TLV, since its value is single `uint8_t` ( (series ID) and we can use the `Tlv` helper methods to read the value directly (we have similar helper methods for writing simple (single-value) TLVs into a message).
     - Something unrelated, I would highly suggest that we add a Tlv prefix to any class that is a TLV and sub-classes `Tlv` type (for example, suggest renaming `LinkMetricsQueryOptions` -> `LinkMetricsQueryOptionsTlv`)
  - For `LinkMetricsQueryOptionsTlv` (which itself can contain an array of `LinkMetricsTypeIdFlags` values, again we cannot know its size and should not assume some max size for it, so we cannot have a buffer and read it whole)
       - We should instead read its value one `TypeIdFlags`at a time and then check if the type id is one of the ones we know and if so, add them into `typeIdFlags[kMaxTypeIdFlags]` array (if not just skip it I guess).

Reply:
Thanks for the very kind explanation.

I have one question here:

> We should instead read its value one TypeIdFlags at a time and then check if the type id is one of the ones we know and if so, add them into typeIdFlags[kMaxTypeIdFlags] array (if not just skip it I guess).

What about reading all `TypeIdFlags` into a buffer once and pass the buffer to `AppendSingleProbeLinkMetricsReport` to filter the ones we know? Because `AppendSingleProbeLinkMetricsReport` would always check the values of `TypeIdFlags`. I think we cannot totally make no assumptions when parsing `LinkMetricsQueryOptions`. Actually I think `kMaxTypeIdFlags` is an assumption about it. Regarding future extension for other types, we can make `kMaxTypeIdFlags` a little larger.

And even we parse the value one by one, it's still possible that the count of `TypeIdFlags` exceeds `kMaxTypeIdFlags` and then we might think this is an invalid Tlv and generate a parse error. So I think we can make assumptions on the length of `LinkMetricsOptionsTlv`, check the length and then we read its value once into a buffer.

Thoughts?

Reply:
> I think we cannot totally make no assumptions when parsing LinkMetricsQueryOptions. Actually I think kMaxTypeIdFlags is an assumption about it. Regarding future extension for other types, we can make kMaxTypeIdFlags a little larger.

Let me describe the scenario where we will face compatibility issues and challenges
- Suppose we use the PR as is and put a hard limit in max typid in the message or in the way the (sub)TLVs are parsed (order/number of sub-TLVs in a message, etc).
- Then say product A is shipped running this implementation.
- It will work fine while all devices are running the same version of sw/OT stack and follow same format.
- Now suppose in next version of thread spec more type ids are defined and/or new sub-TLVs are added.
- We update the OT stack and add support for the new types id and new features.
- Now a product B is shipped running the new version of OT stack.
- We expect product A still running earlier stack version to continue to work with product B running new version (which can now add new sub-TLVs and/or type ids, etc. when sending its requests).
- If we do not implement the parsing/processing of message in a flexible way, we will create a challenge requiring sw updates or patches for all in-field devices for features to continue to work.

-------

> What about reading all TypeIdFlags into a buffer once and pass the buffer to AppendSingleProbeLinkMetricsReport to filter the ones we know?

What buffer size would we use to be safe and compatible all future extensions here? The challenge is that we cannot safely know the size. What we do know currently is which (and how many) type-ids we support in the current implementation. Let me describe my suggestion from above with more details:
- We define a buffer `typeIdFlags[kMaxTypeIdFlags]` with `kMaxTypeIdFlags` (`kMaxTypeIdFlags` is number of type ids we know how to handle and support in current implementation)
- We read the type ids from the TLV and message one by one and if it is one of the ones we know, we add it in `typeIdFlags[]`, if not we just ignore it
    - Basically we read and filter the array of type ids from the message into a subset of type-ids which we know how to handle.
    - If we see duplicate type-id entries, we can consider it invalid (give parse error) or may choose to ignore it (not sure if spec defined any specific behavior - if not specified then I guess it's up to us/developer to decide a reasonable behavior).
- We then pass `typeIdFlags[]` (and its length) to `AppendSingleProbeLinkMetricsReport()`.

Reply:
SGTM. Thanks very much.

Reply:
Updated to add type ids into `typeIdFlags[]` one by one and filter the valid ones.

I also use `otLinkMetrics` for duplicate check. However I prefer to pass `typeIdFlags[]` to `AppendSingleProbeLinkMetricsReport()` as in this function it needs to append the exact same thing in `typeIdFlags[]` to Report Sub-TLV. I think it would be a waste if we convert `otLinkMetrics` to `typeIdFlags[]` again.